### PR TITLE
Make UTS 46 normalization non-experimental

### LINKED ISSUE #2850: Provide UTS 46 mapping / disallow operations fused with normalization
Gecko implements most of its IDNA processing on its own, but it currently uses ICU4C's `uidna_labelToUnicode` for UTS 46 processing.

Code that wraps `ComposingNormalizer::try_new_uts46_without_ignored_and_disallowed_unstable` and augments it to provide the functionality of ICU4C's `uidna_labelToUnicode` needs to go _somewhere_: Since it needs some data for ignored and disallowed that is in sync with ICU4X data, it would make sense to start a new ICU4X component for it. ([We already reserve the name.](https://crates.io/crates/icu_idna))

### PR DESCRIPTION
**Pull request by hsivonen**
* Bake ignored/disallow data into the normalization data after all.
 * Make public operations available via a dedicated wrapper type instead of the main normalizer types.

Closes #2850

### CODE REVIEW

**Comment by hsivonen** [comment]
The ICU PR is https://github.com/unicode-org/icu/pull/2945

**Comment by hsivonen** [comment]
See https://github.com/hsivonen/rust-url/blob/icu4x/idna/src/uts46.rs for usage.

**Comment by hsivonen** [comment]
It seems that the remainin CI failures are due to the ICU4C part not yet contributing to the exported data, so it seems OK to start reviewing this.

It doesn't make much sense to expose this API over FFI. If other languages want ICU4X-backed UTS 46, it makes sense to introduce FFI for https://github.com/hsivonen/rust-url/blob/icu4x/idna/src/uts46.rs .

File: components/normalizer/src/uts46.rs
Comment (hsivonen) [inline components/normalizer/src/uts46.rs:136]: @eggrobin , are you able to / comfortable with guessing at this point which direction the PAG is going with this and opining on whether the special case should remain in this PR?

Comment (hsivonen) [reply components/normalizer/src/uts46.rs:136]: It [seems to me](https://unicode-org.atlassian.net/browse/ICU-22658?focusedCommentId=174822) this special case isn't even needed. Am I missing something?

**Comment by hsivonen** [comment]
Changing review request to @eggrobin per discussion with @echeran .

File: components/normalizer/src/uts46.rs
Comment (eggrobin) [reply components/normalizer/src/uts46.rs:136]: Apologies for the delayed response, I was on vacation.

On where PAG is going: first note that PAG does not decide anything, PAG recommends and the UTC decides.
The planned PAG recommendation includes changing U+2F868, U+2F874, U+2F91F, U+2F95F, and U+2F9BF from disallowed to mapped, but includes other changes as well, trimming the base exclusion set down to 129 characters. The PAG report with these recommendations should be published shortly.

However, there is no decision until UTC #&zwnj;179, on the 23rd to 25th of this month; the UTC does not always follow WG recommendations.

On the implications on your implementation: I do not understand UTS #&zwnj;46 enough to say anything useful, ask @markusicu.

Comment (eggrobin) [reply components/normalizer/src/uts46.rs:136]: See [L2/24-064](https://www.unicode.org/cgi-bin/GetDocumentLink?L2/24-064) Section 6.2.

Comment (hsivonen) [reply components/normalizer/src/uts46.rs:136]: Thank you. Removing the special case.

**Comment by sffc** [comment]
@eggrobin says he is not the correct person to review this from the code side. Adding back @echeran for the code review.

**Review by echeran — DISMISSED** [review]
LGTM

**Comment by echeran** [comment]
question + suggestion (optional): IIUC, at this point, UTS 46 is pretty much IDNA2008. Is that correct? If so, you could make a mention of that in the docs becuase UTS 46 covers not just IDNA2008 but also IDNA2003 and the transition rules between the two. If at this point, everyone has switched over to IDNA2008, including all major browsers, it would help the docs to add that extra specificity.

File: components/normalizer/src/lib.rs
Comment (hsivonen) [reply components/normalizer/src/lib.rs:573]: The underlying struct can be used for non-UTS46 purposes: the usual normalizations. For those, we expose contiguous-buffer APIs, and the contiguous-buffer versions currently can't deal with ignorables, and since there is currently no use case for them to deal with ignorables, I think it's better to have a state that they can `debug_assert!` against than to add support for ignorables is the contiguous-buffer entry points.

**Comment by hsivonen** [comment]
> question + suggestion (optional): IIUC, at this point, UTS 46 is pretty much IDNA2008. Is that correct? If so, you could make a mention of that in the docs becuase UTS 46 covers not just IDNA2008 but also IDNA2003 and the transition rules between the two. If at this point, everyone has switched over to IDNA2008, including all major browsers, it would help the docs to add that extra specificity.

No, UTS 46 non-transitional accepts a superset of what IDNA 2008 accepts and the major browsers use UTS 46 non-transitional specifically. (To use an example from the UTS 46 spec, switching to pure IDNA 2008 would break URLs like http://www.ÖBB.at/ .) Beyond the remark in the docs about the three major browser engines, I'd prefer to leave the characterization of the situation to the UTS 46 spec itself.

> LGTM

Thanks! Since https://github.com/unicode-org/icu/pull/2945 looks ready to land, I'll open another ICU4C PR with a backport to the maintenance branch. Then we can make datagen over on the ICU4X side pull the updated export before merging this.

**Review by sffc — APPROVED** [review]
Re-approving after merge and clippy fix. Replying on @echeran's approval of business logic.

**Pull request merged**
